mlir/test/Target/Cpp/switch.mlir

```diff
@@ -0,0 +1,129 @@
+    %1 = emitc.call_opaque "func_b" () : () -> i32
+    emitc.yield
+  }
+  case 5: {
+    %2 = emitc.call_opaque "func_a" () : () -> i32
+    emitc.yield
+  }
+  default: {
+    %3 = "emitc.variable"(){value = 42.0 : f32} : () -> f32
+    %4 = "emitc.variable"(){value = 42.0 : f32} : () -> f32
+
+    emitc.call_opaque "func2" (%3) : (f32) -> ()
+    emitc.call_opaque "func3" (%3, %4) { args = [1 : index, 0 : index] } : (f32, f32) -> ()
+    emitc.yield
+  }
+  return
+}
+
+// CHECK-LABEL: void emitc_switch_i16() {
+// CHECK:         int16_t v1 = 1;
+// CHECK:         switch(v1) {
+// CHECK:         case (2): {
+// CHECK:           int32_t v2 = func_b();
+// CHECK:           break;
+// CHECK:         }
+// CHECK:         case (5): {
+// CHECK:           int32_t v3 = func_a();
+// CHECK:           break;
+// CHECK:         }
+// CHECK:         default: {
+// CHECK:           float v4 = 4.200000000e+01f;
+// CHECK:           float v5 = 4.200000000e+01f;
+// CHECK:           func2(v4);
+// CHECK:           func3(v5, v4);
+// CHECK:           break;
+// CHECK:         }
+// CHECK:         return;
+// CHECK:       }
+func.func @emitc_switch_i16() {
+  %0 = "emitc.variable"(){value = 1 : i16} : () -> i16
+
+  emitc.switch %0 : i16
+  case 2: {
+    %1 = emitc.call_opaque "func_b" () : () -> i32
+    emitc.yield
+  }
+  case 5: {
+    %2 = emitc.call_opaque "func_a" () : () -> i32
+    emitc.yield
+  }
+  default: {
+    %3 = "emitc.variable"(){value = 42.0 : f32} : () -> f32
+    %4 = "emitc.variable"(){value = 42.0 : f32} : () -> f32
+
+    emitc.call_opaque "func2" (%3) : (f32) -> ()
+    emitc.call_opaque "func3" (%3, %4) { args = [1 : index, 0 : index] } : (f32, f32) -> ()
+    emitc.yield
+  }
+  return
+}
+
+// CHECK-LABEL: void emitc_switch_ui16() {
+// CHECK:         uint16_t v1 = 1;
+// CHECK:         switch(v1) {
+// CHECK:         case (2): {
+// CHECK:           int32_t v2 = func_b();
+// CHECK:           break;
+// CHECK:         }
+// CHECK:         case (5): {
+// CHECK:           int32_t v3 = func_a();
+// CHECK:           break;
+// CHECK:         }
+// CHECK:         default: {
+// CHECK:           float v4 = 4.200000000e+01f;
+// CHECK:           float v5 = 4.200000000e+01f;
+// CHECK:           func2(v4);
+// CHECK:           func3(v5, v4);
+// CHECK:           break;
+// CHECK:         }
+// CHECK:         return;
+// CHECK:       }
+func.func @emitc_switch_ui16() {
+  %0 = "emitc.variable"(){value = 1 : ui16} : () -> ui16
+
+  emitc.switch %0 : ui16
+  case 2: {
+    %1 = emitc.call_opaque "func_b" () : () -> i32
+    emitc.yield
+  }
+  case 5: {
+    %2 = emitc.call_opaque "func_a" () : () -> i32
+    emitc.yield
+  }
+  default: {
+    %3 = "emitc.variable"(){value = 42.0 : f32} : () -> f32
+    %4 = "emitc.variable"(){value = 42.0 : f32} : () -> f32
+
+    emitc.call_opaque "func2" (%3) : (f32) -> ()
+    emitc.call_opaque "func3" (%3, %4) { args = [1 : index, 0 : index] } : (f32, f32) -> ()
+    emitc.yield
+  }
+  return
+}
```

Comment:
The documentation for emitc.yield says `If the parent operation defines no values, then the `emitc.yield` may be left out in the custom syntax and the builders will insert one implicitly.`
Does that work here? If that would work, I think it would be easy to confuse with fall-through, which this PR doesn't support.

Reply:
Right now, you must explicitly yield each case region and the default one. We can support the implicit addition of `emitc::YieldOp` to the end of each region in the parser, as it is done in `emitc::IfOp`'s parser with `IfOp::ensureTerminator(...)`.

Reply:
Ok, then I think it's better to continue requiring explicit `yields` for now so users don't get confused about a missing yield and think that this is doing fall-through.

Reply:
> Ok, then I think it's better to continue requiring explicit yields for now so users don't get confused about a missing yield and think that this is doing fall-through.

I will add `requiring explicit yields` to the description of the op in the `.td` file. Can I resolve conversation?

Reply:
yes, that sounds good